dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/TrackerObjectsMapper.java

```diff
@@ -0,0 +1,348 @@
+    dbEvent.setOccurredDate(DateUtils.fromInstant(event.getOccurredAt()));
+    dbEvent.setScheduledDate(DateUtils.fromInstant(event.getScheduledAt()));
+
+    if (event.getAttributeOptionCombo().isNotBlank()) {
+      dbEvent.setAttributeOptionCombo(
+          preheat.getCategoryOptionCombo(event.getAttributeOptionCombo()));
+    } else {
+      dbEvent.setAttributeOptionCombo(preheat.getDefault(CategoryOptionCombo.class));
+    }
+
+    dbEvent.setGeometry(event.getGeometry());
+
+    EventStatus currentStatus = event.getStatus();
+    EventStatus previousStatus = dbEvent.getStatus();
+
+    if (currentStatus != previousStatus && currentStatus == EventStatus.COMPLETED) {
+      dbEvent.setCompletedDate(now);
+      dbEvent.setCompletedBy(user.getUsername());
+    }
+
+    if (currentStatus != EventStatus.COMPLETED) {
+      dbEvent.setCompletedDate(null);
+      dbEvent.setCompletedBy(null);
+    }
+
+    dbEvent.setStatus(currentStatus);
+
+    if (Boolean.TRUE.equals(programStage.isEnableUserAssignment())
+        && event.getAssignedUser() != null
+        && !event.getAssignedUser().isEmpty()) {
+      Optional<User> assignedUser =
+          preheat.getUserByUsername(event.getAssignedUser().getUsername());
+      assignedUser.ifPresent(dbEvent::setAssignedUser);
+    }
+
+    if (program.isRegistration()
+        && dbEvent.getScheduledDate() == null
+        && dbEvent.getOccurredDate() != null) {
+      dbEvent.setScheduledDate(dbEvent.getOccurredDate());
+    }
+
+    // TODO(DHIS2-18223): Remove data value mapping and fix changelog logic
+    for (DataValue dataValue : event.getDataValues()) {
+      DataElement dataElement = preheat.getDataElement(dataValue.getDataElement());
+
+      EventDataValue eventDataValue = new EventDataValue();
+      eventDataValue.setDataElement(dataElement.getUid());
+      eventDataValue.setCreated(DateUtils.fromInstant(dataValue.getCreatedAt()));
+      eventDataValue.setCreatedByUserInfo(UserInfoSnapshot.from(user));
+      eventDataValue.setValue(dataValue.getValue());
+      eventDataValue.setLastUpdated(now);
+      eventDataValue.setProvidedElsewhere(dataValue.isProvidedElsewhere());
+      eventDataValue.setLastUpdatedByUserInfo(UserInfoSnapshot.from(user));
+
+      dbEvent.getEventDataValues().add(eventDataValue);
+    }
+
+    if (isNotEmpty(event.getNotes())) {
+      dbEvent
+          .getNotes()
+          .addAll(
+              event.getNotes().stream()
+                  .map(note -> map(preheat, note, user))
+                  .collect(Collectors.toSet()));
+    }
+
+    return dbEvent;
+  }
+
+  public static @Nonnull Relationship map(
+      @Nonnull TrackerPreheat preheat,
+      @Nonnull org.hisp.dhis.tracker.imports.domain.Relationship relationship,
+      @Nonnull UserDetails user) {
+    RelationshipType relationshipType =
+        preheat.getRelationshipType(relationship.getRelationshipType());
+    RelationshipItem fromItem = new org.hisp.dhis.relationship.RelationshipItem();
```

Comment:
declare vars as close to their usage. easier to read and added benefit of sometimes narrowing their scope